### ISSUE TITLE
Regression: Conflicts no longer open properly in kdiff3

### ISSUE THREAD

**Issue by jansennerd10** [gitextensions/gitextensions#7544]
## Current behaviour

After clicking "Merge" in the Resolve merge conflicts window, kdiff3 opens with no files selected.

## Expected behaviour

kdiff3 opens with the base, local, and remote versions of the conflicted file.

## Steps to reproduce

Start with GitExtensions v3.2 installed; upgrade to v3.3 using the built-in upgrade functionality; try to perform a merge between branches which have conflicts. Choose to resolve conflicts; select a conflicted file in the Resolve merge conflicts window; then click Merge.

I have not evaluated whether this applies when GitExtensions v3.3 is installed from scratch, as either way this is a bug. However, I have reverted to v3.2, observed this working properly, then upgraded again and observed the same problem.

## Screenshots

kdiff3 window opened by GitExtensions v3.3
![image](https://user-images.githubusercontent.com/15092515/71301231-4afbbb00-2362-11ea-94d8-1105f9d3d250.png)

## Did this work in previous version of GitExtensions

Yes, in v3.2.

## Environment

- Git Extensions 3.3.0.7719
- Build e52f5178e8d3c9899efbe6fae66026865afa3fe1
- Git 2.23.0.windows.1
- Microsoft Windows NT 10.0.17763.0
- .NET Framework 4.7.3460.0
- DPI 192dpi (200% scaling)

## Diagnostics

**Comment by jansennerd10** [comment]
From the command log in v3.2:
```
C:\Program Files\KDiff3\kdiff3.exe "C:\GitRepos\my-repo\UnitTestProject1/UnitTest1.cs.BASE" "C:\GitRepos\my-repo\UnitTestProject1/UnitTest1.cs.LOCAL" "C:\GitRepos\my-repo\UnitTestProject1/UnitTest1.cs.REMOTE" -o "UnitTestProject1/UnitTest1.cs"
```

From the command log in v3.3:
```
C:\Program Files\KDiff3\kdiff3.exe
```

**Comment by jrfnl** [comment]
I can confirm this is happening. Running into the same issue myself.

**Comment by RussKie** [comment]
Can you please share screenshots of your _Settings > Git > Config_ configs?

**Comment by RussKie** [comment]
@gerhardol could this be a result of #7009?

**Comment by jrfnl** [comment]
*grin* I'd literally just made one.
![image](https://user-images.githubusercontent.com/663378/71307018-a5d7f580-23e8-11ea-84d0-6fa58b4af0c6.png)

**Comment by jrfnl** [comment]
Extra info:

* Clicking `Suggest` next to `Mergetool command` has no effect.
* And as the line is greyed, there's no way to edit it either.

Tried to re-install 3.2.1, but the `msi` installer blocks this with an "newer version already installed" message.
I seem to remember that didn't use to happen as I've had to revert back to older versions a few times before.

![image](https://user-images.githubusercontent.com/663378/71307116-cf455100-23e9-11ea-8479-7ecf954e6cb3.png)

**Comment by jrfnl** [comment]
Hmm.. noticed the following while making a backup of my settings in `%user%/AppData/Roaming` (before doing a hard uninstall and attempting another re-install of 3.2.1):

Guess the settings directory name was changed without an upgrade routine associated with it ?

![image](https://user-images.githubusercontent.com/663378/71307173-c012d300-23ea-11ea-88c9-7139203967fc.png)

**Comment by jrfnl** [comment]
I'm not having a good day. Running into problems re-installing 3.2.1 with the below error - while I am the system administrator....:

![image](https://user-images.githubusercontent.com/663378/71307385-fe5dc180-23ed-11ea-9de5-30b28ca67877.png)

**Comment by RussKie** [comment]
Run as admin?

**Comment by jrfnl** [comment]
@RussKie For some mysterious reason that option doesn't show on right-click of a `msi` file (it does for everything else).

Hacking via regedit now to see if I can force it to run as admin anyway,.

**Comment by jrfnl** [comment]
Tried everything I could think of, but can't get v 3.2.1 to reinstall. When run as administrator I get this and clicking ok exits the program:

![image](https://user-images.githubusercontent.com/663378/71307715-c907a280-23f2-11ea-8940-8d7670a1b5c0.png)

**Comment by jrfnl** [comment]
I even get the "you need admin privileges" message for the "Install just for you / You do not need local Administrator privileges" option... *sigh* Sometimes I **so** hate Windows.

**Comment by jrfnl** [comment]
Ok, so I had to go all the way back to `GitExtensions-3.1.0.5877.msi` before I could get it to install without the privilege error. But trying to upgrade from there is still running into the same issues.
Going to try again after a system restart.

I do wonder what was changed in the installation package between then and now which is causing these issues.....

**Comment by RussKie** [comment]
You could also fix the issue by manually tweaking .gitconfig:

```
[merge]
	tool = kdiff3
[diff]
	guitool = kdiff3
[difftool "kdiff3"]
	path = C:/path/to/KDiff3/KDiff3.exe
	cmd = \"C:/path/to/KDiff3/KDiff3.exe\" \"$LOCAL\" \"$REMOTE\"
[mergetool "kdiff3"]
	path = C:/path/to/KDiff3/KDiff3.exe
	cmd = \"C:/path/to/KDiff3/KDiff3.exe\" \"$LOCAL\" \"$REMOTE\" \"$BASE\" \"$MERGED\"
```

**Comment by jrfnl** [comment]
@RussKie Will try. But I first need to be able to get a more recent GE version installed again & working at all...

**Comment by RussKie** [comment]
You can just download on from GitHub, you don't need to auto-update through
the app.
Or use chocolatey

**Comment by jrfnl** [comment]
@RussKie I've been using the downloads from GH. Those were where I ran into all the install problems with...

**Comment by jrfnl** [comment]
Ok, so after a cold reboot of the system I managed to get GE installed again.

@RussKie I've manually edited the `.gitconfig` file as you suggested and set the command. I've got a very large rebase to do, so I will no doubt run into conflicts before too long, so I will let you know whether it worked.

**Comment by gerhardol** [comment]
> @gerhardol could this be a result of #7009?

That PR is the major change here, so likely involved. kdiff3 is handled as all other difftools now, as the options are there, I cannot see what the problem is.

**Comment by jrfnl** [comment]
@RussKie Ok, so I've tried with the command in `.gitconfig`, but that didn't have the correct parameters.
* The order of the files was different than before (and illogical)
* The "merge" window didn't show up.

After going through the [docs of KDiff3](http://kdiff3.sourceforge.net/doc/documentation.html#commandline), I got it working as it was in v 3.2.1 with the following settings manually added to the `.gitconfig` file:
```
[mergetool "kdiff3"]
	path = C:/path/to/KDiff3/kdiff3.exe
	cmd = \"C:/path/to/KDiff3/kdiff3.exe\" \"$BASE\" \"$LOCAL\" \"$REMOTE\" -o \"$MERGED\"
```

**Comment by jrfnl** [comment]
> I cannot see what the problem is.

The problem is that something which was working perfectly fine in previous versions of GE and is a quite crucial feature, suddenly stopped working completely.

The above is a manual work-around which will get it working again, but this is:
1. Not obvious.
2. A breaking change for anyone using Kdiff3 - which for the longest time was the default and included with the GE installer.
3. Not even mentioned in the changelog.

Just a suggestion, but if something this crucial changes, it may help to have an upgrade routine which will check whether `kdiff3` is set as the diff/merge tool for the user and if so, automatically make the change to the `.gitconfig` file.

**Comment by mstv** [comment]
Selecting `Mergetool` "kdiff3" disables the input field `Mergetool command` -- even with the one year old
- Git Extensions **3.00.00.4000Dev@cf9ca**
- Git **2.24.1.windows.2**
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.4075.0

Could it depend also on the git version?

**Comment by jrfnl** [comment]
I'm on Git 2.24.1.windows.2 as well, but already was before I upgraded GE and things were working fine before the upgrade.

**Comment by MarkJ74** [comment]
I ran into this as well, I uninstalled GE 3.2.1 and installed 3.3, then noticed that kdiff didn't work with merge conflicts, so I uninstalled 3.3 and put 3.2.1 back on, and it worked ok again.

**Comment by gerhardol** [comment]
I have submitted PRs (both master and 3.3) for this issue.
There are two workarounds:
 * Set mergetool cmd separately. Use the portable build in #7549, command line or whatever.
 * Only use Start Mergetool to start merge conflicts with files

There are basically two implementations of of starting mergetools, I propose to remove the problematic in #7550

In 2.x, kdiff whas bundled and there were various hardcoding to handle the mergetools. The bundling was removed but some of the unneeded magic remained, removed in 3.3 (#7009). The handling has already been cleaned up in master (#7544), also allowing manual changes to known difftools. In #7009, the special handling of kdiff3 in one of the ways invoking it was missed, why this occurred.
